### PR TITLE
Optimize scalar and scalar pair representations loaded from ByRef in llvm

### PR DESCRIPTION
**Pull request by oli-obk**
in https://github.com/rust-lang/rust/pull/105653 I noticed that we were generating suboptimal LLVM IR if we had a `ConstValue::ByRef` that could be represented by a `ScalarPair`. Before https://github.com/rust-lang/rust/pull/105653 this is probably rare, but after it, every slice will go down this suboptimal code path that requires LLVM to untangle a bunch of indirections and translate static allocations that are only used once to read a scalar pair from.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/builder.rs [hunk 1]
```diff
@@ -1233,6 +1235,92 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
...
+                        self,
+                        alloc_range(Size::ZERO, a_size),
+                        /*read_provenance*/ matches!(a.primitive(), abi::Pointer(_)),
+                    )
+                    .unwrap();
+                let a_val = self.scalar_to_backend(
+                    a_val,
+                    a,
+                    self.scalar_pair_element_backend_type(layout, 0, true),
+                );
```
Comment (cjgillot) [inline compiler/rustc_codegen_llvm/src/builder.rs:1289]: This snippet appear 3 times. Does it deserve a helper closure?

File: compiler/rustc_codegen_gcc/src/builder.rs [hunk 2]
```diff
@@ -1425,6 +1428,23 @@ impl<'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
...
+        else {
+            let init = const_alloc_to_gcc(self, alloc);
+            let base_addr = self.static_addr_of(init, alloc.inner().align, None);
+
+            let array = self.const_bitcast(base_addr, self.type_i8p());
+            let value = self.context.new_array_access(None, array, self.const_usize(offset.bytes())).get_address(None);
+            let value = self.const_bitcast(value, ty);
+            self.load_operand(PlaceRef::new_sized(value, layout))
+        }
+    }
```
Comment (cjgillot) [inline compiler/rustc_codegen_gcc/src/builder.rs:1447]: codegen_gcc is updated. Why isn't codegen_cranelift? It doesn't use that trait?

File: compiler/rustc_codegen_llvm/src/builder.rs [hunk 3]
```diff
@@ -1233,6 +1235,92 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
...
+        let alloc_align = alloc.inner().align;
+        assert_eq!(alloc_align, layout.align.abi);
+        let llty = self.type_ptr_to(layout.llvm_type(self));
+        // It may seem like all types with `Scalar` or `ScalarPair` ABI are fair game at this point.
+        // However, `MaybeUninit<u64>` is considered a `Scalar` as far as its layout is concerned --
+        // and yet cannot be represented by an interpreter `Scalar`, since we have to handle the
+        // case where some of the bytes are initialized and others are not. So, we need an extra
+        // check that walks over the type of `mplace` to make sure it is truly correct to treat this
+        // like a `Scalar` (or `ScalarPair`).
+        match layout.abi {
```
Comment (cjgillot) [inline compiler/rustc_codegen_llvm/src/builder.rs:1255]: Why can't we handle these cases in `codegen_ssa`.

**Review by cjgillot — COMMENTED** [review]
Could you add a codegen test?

File: compiler/rustc_codegen_gcc/src/builder.rs (hunk 2 above)
Comment (bjorn3) [reply compiler/rustc_codegen_gcc/src/builder.rs:1447]: It doesn't.

**Comment by oli-obk** [comment]
Addressed all review comments and added a codegen test (before the commit that changes things, so you can see the behaviour change to the test on the second commit)

**Comment by cjgillot** [comment]
Thanks.

Should we eventually have a `ConstValue::ScalarPair`, or does that go against the design you are aiming for?

In the mean time:
@bors r+

**Comment by oli-obk** [comment]
> Should we eventually have a `ConstValue::ScalarPair`, or does that go against the design you are aiming for?

I'm trying to remove that, and instead just rely on `ByRef` if we can get acceptable performance this way

**Pull request merged**